diff --git a/Makefile b/Makefile
index a61a95b6b38f..16be786d29e3 100644
--- a/Makefile
+++ b/Makefile
@@ -698,8 +698,12 @@ KBUILD_CFLAGS      += $(call cc-disable-warning, address-of-packed-member)
 ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS  += -Os
 else
+ifdef CONFIG_CC_OPTIMIZE_HARDER
+KBUILD_CFLAGS  += -O3 $(call cc-disable-warning,maybe-uninitialized,)
+else
 KBUILD_CFLAGS   += -O2
 endif
+endif

 ifdef CONFIG_CC_DISABLE_WARN_MAYBE_UNINITIALIZED
 KBUILD_CFLAGS   += -Wno-maybe-uninitialized
diff --git a/init/Kconfig b/init/Kconfig
index 8b9ffe236e4f..7bf517949d32 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1137,6 +1137,13 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
          with the "-O2" compiler flag for best performance and most
          helpful compile-time warnings.

+config CC_OPTIMIZE_HARDER
+       bool "Optimize harder"
+       help
+         This option will pass "-O3" to your compiler resulting in a
+         larger and faster kernel. The more complex optimizations also
+         increase compilation time and may affect stability.
+
 config CC_OPTIMIZE_FOR_SIZE
        bool "Optimize for size"
        imply CC_DISABLE_WARN_MAYBE_UNINITIALIZED  # avoid false positives
